go-controller/pkg/allocator/pod/pod_annotation_test.go

```diff
@@ -575,6 +580,130 @@ func Test_allocatePodAnnotationWithRollback(t *testing.T) {
+			ipam:                            true,
+			enablePreconfiguredUDNAddresses: true,
+			role:                            types.NetworkRolePrimary, // has to be primary network for default routes to be set
+			netConf: &ovncnitypes.NetConf{
+				Topology: types.Layer2Topology,
+				NetConf: cnitypes.NetConf{
+					Name: "network",
+				},
+				NADName: util.GetNADName("namespace", "network"),
+				Subnets: "192.168.0.0/24,2001:db8::/64",
+				Role:    types.NetworkRolePrimary,
+			},
+			args: args{
+				network: &nadapi.NetworkSelectionElement{
+					MacRequest: requestedMAC,
+					IPRequest:  []string{"192.168.0.101/24"},
+				},
+				ipAllocator: &ipAllocatorStub{
+					nextIPs: ovntest.MustParseIPNets("192.168.0.3/24"),
+				},
+			},
+			wantUpdatedPod: true,
+			wantPodAnnotation: &util.PodAnnotation{
+				IPs:      ovntest.MustParseIPNets("192.168.0.101/24"),
+				MAC:      requestedMACParsed,
+				Gateways: []net.IP{ovntest.MustParseIP("192.168.0.1").To4()},
+				Routes: []util.PodRoute{
+					{
+						Dest:    ovntest.MustParseIPNet("100.65.0.0/16"),
+						NextHop: ovntest.MustParseIP("192.168.0.1").To4(),
+					},
+				},
+				Role: types.NetworkRolePrimary,
+			},
+			wantReleasedIPsOnRollback: ovntest.MustParseIPNets("192.168.0.101/24"),
+		},
+		{
+			// on networks with IPAM and layer2 topology, expect error when EnablePreconfiguredUDNAddresses is false
+			name: "expect error, static IP with IPAM on layer2 when EnablePreconfiguredUDNAddresses is false",
+			ipam: true,
+			// enablePreconfiguredUDNAddresses defaults to false
+			netConf: &ovncnitypes.NetConf{
+				Topology: types.Layer2Topology,
+				NetConf: cnitypes.NetConf{
+					Name: "network",
+				},
+				NADName: util.GetNADName("namespace", "network"),
+				Subnets: "192.168.0.0/24,2001:db8::/64",
+			},
+			args: args{
+				network: &nadapi.NetworkSelectionElement{
+					IPRequest: []string{"192.168.0.101/24"},
+				},
+			},
+			wantErr: true,
+		},
+		{
+			// with preconfigured UDN address feature enabled still continue failing with secondary layer2 with ipam + static IPs
+			name:                            "expect error, static IP with IPAM on secondary network when EnablePreconfiguredUDNAddresses is enabled",
+			ipam:                            true,
+			enablePreconfiguredUDNAddresses: true,
+			netConf: &ovncnitypes.NetConf{
+				Topology: types.Layer2Topology,
+				NetConf: cnitypes.NetConf{
+					Name: "network",
+				},
+				NADName: util.GetNADName("namespace", "network"),
+				Subnets: "192.168.0.0/24,2001:db8::/64",
+				Role:    types.NetworkRoleSecondary, // This makes it a secondary network
+			},
+			args: args{
+				network: &nadapi.NetworkSelectionElement{
+					IPRequest: []string{"192.168.0.101/24"},
+				},
+			},
+			role:    types.NetworkRoleSecondary,
+			wantErr: true,
+		},
+		{
+			// on primary networks with IPAM and layer2 topology, expect success when EnablePreconfiguredUDNAddresses is enabled
```

Comment:
I am sooooo confused. This has too many parameters, making it hard to follow. Or I'm daft. 

Isn't this the same test you have added on line 585 ?

Reply:
could we kind of tinker with the test names or comments to reflect that ? 

honestly, I get headaches from looking at these test entries.